perf/evp_fetch.c

```diff
@@ -0,0 +1,272 @@
+            if (cph == NULL) {
+                fprintf(stderr, "Failed to fetch %s\n", fetch_alg);
+                err = 1;
+                return;
+            }
+            EVP_CIPHER_free(cph);
+            break;
+        case FETCH_KDF:
+            EVP_KDF *kdf = EVP_KDF_fetch(ctx, fetch_alg,
+                                         fetch_entries[j].propq);
+            if (kdf == NULL) {
+                fprintf(stderr, "Failed to fetch %s\n", fetch_alg);
+                err = 1;
+                return;
+            }
+            EVP_KDF_free(kdf);
+            break;
+        case FETCH_MAC:
+            EVP_MAC *mac = EVP_MAC_fetch(ctx, fetch_alg,
+                                         fetch_entries[j].propq);
+            if (mac == NULL) {
+                fprintf(stderr, "Failed to fetch %s\n", fetch_alg);
+                err = 1;
+                return;
+            }
+            EVP_MAC_free(mac);
+            break;
+        case FETCH_RAND:
+            EVP_RAND *rnd = EVP_RAND_fetch(ctx, fetch_alg,
+                                           fetch_entries[j].propq);
+            if (rnd == NULL) {
+                fprintf(stderr, "Failed to fetch %s\n", fetch_alg);
+                err = 1;
+                return;
+            }
+            EVP_RAND_free(rnd);
+            break;
+        default:
+            err = 1;
+            return;
+        }
+    }
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+}
+
+int main(int argc, char *argv[])
+{
+    OSSL_TIME duration;
+    OSSL_TIME ttime;
+    double real_num_calls;
+    double av;
+    int terse = 0;
+    int argnext;
+    size_t i;
+    int rc = EXIT_FAILURE;
+    char *fetch_type = getenv("EVP_FETCH_TYPE");
+
+    if ((argc != 2 && argc != 3)
+                || (argc == 3 && strcmp("--terse", argv[1]) != 0)) {
+        printf("Usage: %s [--terse] threadcount\n", argv[0]);
+        return EXIT_FAILURE;
+    }
+
+    if (argc == 3) {
+        terse = 1;
+        argnext = 2;
+    } else {
+        argnext = 1;
+    }
+
+    if (fetch_type != NULL) {
+        exclusive_fetch_alg = strstr(fetch_type, ":");
+        if (exclusive_fetch_alg == NULL) {
+            printf("Malformed EVP_FETCH_TYPE TYPE:ALG\n");
+            return EXIT_FAILURE;
+        }
+        /* Split the string into a type and alg */
+        *exclusive_fetch_alg = '\0';
+        exclusive_fetch_alg++;
+        for (i = 0; i < ARRAY_SIZE(type_map); i++) {
+            if (!strcmp(fetch_type, type_map[i].name)) {
+                exclusive_fetch_type = type_map[i].id;
+                break;
+            }
+        }
+        if (i == ARRAY_SIZE(type_map)) {
+            printf("EVP_FETCH_TYPE is invalid\n");
+            return EXIT_FAILURE;
+        }
+    }
+
+    threadcount = atoi(argv[argnext]);
+    if (threadcount < 1) {
+        printf("threadcount must be > 0\n");
+        return EXIT_FAILURE;
+    }
+
+    ctx = OSSL_LIB_CTX_new();
```

Comment:
I wonder if we should have a "warm up" option that would do one fetch here outside the measurements. The initial fetch on a libctx with no providers loaded will load the default provider implicitly and fighting for that could skew the numbers fairly visibly. I.e. it would be interesting to see the differences in the numbers. Or instead we could explicitly load the default provider with OSSL_PROVIDER_load().

Reply:
I suppose that would depend on the case we want to measure.  I could envision wanting to measure both the fast path (where a fetch would always hit in the cache), and the slow path (where a fetch would never hit in the cache)

Reply:
Another PR I think.

IMO a warm up isn't going to impact the measurement significantly given a large number of iterations per thread.